File: iosApp/iosApp/Components/CommentSheet.swift
import SwiftUI

struct CommentSheet: View {
    let postId: String
    var onDismiss: () -> Void

    @Environment(\.appColors) private var colors
    @State private var comments: [ApiComment] = []
    @State private var isLoading = true
    @State private var commentText = ""
    @State private var replyingTo: ApiComment?
    @State private var isSending = false

    var body: some View {
        VStack(spacing: 0) {
            // Header
            HStack {
                Text("Comentários")
                    .font(.system(size: 17, weight: .semibold))
                    .foregroundColor(colors.textPrimary)
                Spacer()
                Button(action: onDismiss) {
                    Image(systemName: "xmark")
                        .font(.system(size: 14, weight: .medium))
                        .foregroundColor(colors.textSecondary)
                        .frame(width: 28, height: 28)
                }
                .buttonStyle(.plain)
            }
            .padding(.horizontal, 20)
            .padding(.vertical, 16)

            Rectangle().fill(colors.divider).frame(height: 0.5)

            // Content
            if isLoading {
                Spacer()
                ProgressView().tint(colors.accent)
                Spacer()
            } else if comments.isEmpty {
                Spacer()
                Text("Nenhum comentário ainda")
                    .font(.system(size: 14))
                    .foregroundColor(colors.textSecondary)
                Spacer()
            } else {
                ScrollView {
                    LazyVStack(spacing: 0) {
                        ForEach(comments) { comment in
                            CommentItemView(
                                comment: comment,
                                onReply: { replyingTo = $0 },
                                onVote: { id, type in vote(id, type) },
                                onDelete: { id in deleteComment(id) }
                            )
                        }
                    }
                }
            }

            // Reply indicator
            if let reply = replyingTo {
                HStack {
                    Text("Respondendo a \(reply.authorName)")
                        .font(.system(size: 12))
                        .foregroundColor(colors.accent)
                    Spacer()
                    Button { replyingTo = nil } label: {
                        Image(systemName: "xmark")
                            .font(.system(size: 10))
                            .foregroundColor(colors.textSecondary)
                    }
                    .buttonStyle(.plain)
                }
                .padding(.horizontal, 20)
                .padding(.vertical, 8)
                .background(colors.surface)
            }

            // Input
            Rectangle().fill(colors.divider).frame(height: 0.5)
            HStack(spacing: 8) {
                TextField("Adicionar comentário...", text: $commentText)
                    .font(.system(size: 14))
                    .foregroundColor(colors.textPrimary)
                    .padding(.horizontal, 16)
                    .padding(.vertical, 10)
                    .background(colors.surface)
                    .clipShape(RoundedRectangle(cornerRadius: 20))

                Button(action: sendComment) {
                    if isSending {
                        ProgressView()
                            .tint(colors.accent)
                            .frame(width: 36, height: 36)
                    } else {
                        Image(systemName: "paperplane.fill")
                            .font(.system(size: 16))
                            .foregroundColor(commentText.isEmpty ? colors.textSecondary : colors.accent)
                            .frame(width: 36, height: 36)
                    }
                }
                .buttonStyle(.plain)
                .disabled(commentText.isEmpty || isSending)
            }
            .padding(.horizontal, 16)
            .padding(.vertical, 12)
        }
        .background(colors.background)
        .task { await loadComments() }
    }

    private func loadComments() async {
        isLoading = true
        do {
            comments = try await ApiClient.shared.getComments(postId: postId)
        } catch {}
        isLoading = false
    }

    private func sendComment() {
        guard !commentText.isEmpty, !isSending else { return }
        let text = commentText
        let parentId = replyingTo?.id
        commentText = ""
        replyingTo = nil
        isSending = true

        Task {
            do {
                _ = try await ApiClient.shared.addComment(
                    postId: postId,
                    text: text,
                    authorName: TokenManager.shared.displayName ?? "Anônimo",
                    parentId: parentId
                )
                await loadComments()
            } catch {}
            isSending = false
        }
    }

    private func vote(_ commentId: String, _ type: String) {
        Task {
            try? await ApiClient.shared.voteComment(commentId: commentId, voteType: type)
            await loadComments()
        }
    }

    private func deleteComment(_ commentId: String) {
        Task {
            try? await ApiClient.shared.deleteComment(commentId: commentId)
            await loadComments()
        }
    }
}

// MARK: - Comment Item

private struct CommentItemView: View {
    let comment: ApiComment
    var onReply: (ApiComment) -> Void
    var onVote: (String, String) -> Void
    var onDelete: (String) -> Void
    var isReply: Bool = false

    @Environment(\.appColors) private var colors

    var body: some View {
        VStack(alignment: .leading, spacing: 4) {
            // Author + time
            HStack(spacing: 6) {
                Text(comment.authorName.isEmpty ? "Anônimo" : comment.authorName)
                    .font(.system(size: 13, weight: .semibold))
                    .foregroundColor(colors.textPrimary)
                Text("·").foregroundColor(colors.textSecondary)
                Text(comment.timeAgo)
                    .font(.system(size: 11))
                    .foregroundColor(colors.textSecondary)
            }

            Text(comment.text)
                .font(.system(size: 14))
                .foregroundColor(colors.textPrimary)
                .lineSpacing(2)

            // Actions
            HStack(spacing: 4) {
                Button { onVote(comment.id, "up") } label: {
                    HStack(spacing: 2) {
                        Image(systemName: "chevron.up")
                            .font(.system(size: 12))
                        if comment.upvotes > 0 {
                            Text("\(comment.upvotes)").font(.system(size: 11))
                        }
                    }
                    .foregroundColor(colors.textSecondary)
                    .padding(4)
                }
                .buttonStyle(.plain)

                Button { onVote(comment.id, "down") } label: {
                    HStack(spacing: 2) {
                        Image(systemName: "chevron.down")
                            .font(.system(size: 12))
                        if comment.downvotes > 0 {
                            Text("\(comment.downvotes)").font(.system(size: 11))
                        }
                    }
                    .foregroundColor(colors.textSecondary)
                    .padding(4)
                }
                .buttonStyle(.plain)

                if !isReply {
                    Button { onReply(comment) } label: {
                        Text("Responder")
                            .font(.system(size: 12, weight: .medium))
                            .foregroundColor(colors.textSecondary)
                            .padding(4)
                    }
                    .buttonStyle(.plain)
                }

                Spacer()

                Button { onDelete(comment.id) } label: {
                    Image(systemName: "trash")
                        .font(.system(size: 12))
                        .foregroundColor(colors.textSecondary.opacity(0.5))
                }
                .buttonStyle(.plain)
            }
            .padding(.top, 2)

            // Replies
            if !comment.replies.isEmpty {
                ForEach(comment.replies) { reply in
                    CommentItemView(
                        comment: reply,
                        onReply: onReply,
                        onVote: onVote,
                        onDelete: onDelete,
                        isReply: true
                    )
                    .padding(.leading, 20)
                }
            }

            if !isReply {
                Rectangle()
                    .fill(colors.divider.opacity(0.5))
                    .frame(height: 0.5)
                    .padding(.top, 8)
            }
        }
        .padding(.horizontal, isReply ? 0 : 20)
        .padding(.vertical, isReply ? 4 : 12)
    }
}

File: androidApp/src/main/kotlin/com/example/app/android/components/CommentSheet.kt
package com.example.app.android.components

import androidx.compose.animation.animateContentSize
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.imePadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.Send
import androidx.compose.material.icons.filled.Close
import androidx.compose.material.icons.filled.Delete
import androidx.compose.material.icons.filled.KeyboardArrowDown
import androidx.compose.material.icons.filled.KeyboardArrowUp
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.example.app.android.network.ApiClient
import com.example.app.android.network.TokenManager
import com.example.app.android.network.models.Comment
import com.example.app.android.theme.AppTheme
import kotlinx.coroutines.launch

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun CommentSheet(
    postId: String,
    onDismiss: () -> Unit
) {
    val colors = AppTheme.colors
    val scope = rememberCoroutineScope()
    val sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true)

    var comments by remember { mutableStateOf<List<Comment>>(emptyList()) }
    var isLoading by remember { mutableStateOf(true) }
    var commentText by remember { mutableStateOf("") }
    var replyingTo by remember { mutableStateOf<Comment?>(null) }
    var isSending by remember { mutableStateOf(false) }

    LaunchedEffect(postId) {
        isLoading = true
        ApiClient.getComments(postId).fold(
            onSuccess = { comments = it; isLoading = false },
            onFailure = { isLoading = false }
        )
    }

    fun refreshComments() {
        scope.launch {
            ApiClient.getComments(postId).onSuccess { comments = it }
        }
    }

    ModalBottomSheet(
        onDismissRequest = onDismiss,
        sheetState = sheetState,
        containerColor = colors.background,
        dragHandle = null,
        modifier = Modifier.fillMaxHeight(0.85f)
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .imePadding()
        ) {
            // Header
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(horizontal = 20.dp, vertical = 16.dp),
                verticalAlignment = Alignment.CenterVertically
            ) {
                Text(
                    text = "Coment\u00e1rios",
                    fontSize = 17.sp,
                    fontWeight = FontWeight.SemiBold,
                    color = colors.textPrimary,
                    modifier = Modifier.weight(1f)
                )
                IconButton(onClick = onDismiss, modifier = Modifier.size(28.dp)) {
                    Icon(Icons.Default.Close, "Fechar", tint = colors.textSecondary, modifier = Modifier.size(20.dp))
                }
            }

            Box(Modifier.fillMaxWidth().height(0.5.dp).background(colors.divider))

            // Content
            if (isLoading) {
                Box(Modifier.weight(1f).fillMaxWidth(), contentAlignment = Alignment.Center) {
                    CircularProgressIndicator(color = colors.accent, strokeWidth = 2.dp, modifier = Modifier.size(24.dp))
                }
            } else if (comments.isEmpty()) {
                Box(Modifier.weight(1f).fillMaxWidth(), contentAlignment = Alignment.Center) {
                    Text("Nenhum coment\u00e1rio ainda", fontSize = 14.sp, color = colors.textSecondary)
                }
            } else {
                LazyColumn(
                    modifier = Modifier.weight(1f).fillMaxWidth(),
                    verticalArrangement = Arrangement.spacedBy(0.dp)
                ) {
                    items(comments, key = { it.id }) { comment ->
                        CommentItem(
                            comment = comment,
                            onReply = { replyingTo = it },
                            onVote = { id, type ->
                                scope.launch {
                                    ApiClient.voteComment(id, type)
                                    refreshComments()
                                }
                            },
                            onDelete = { id ->
                                scope.launch {
                                    ApiClient.deleteComment(id)
                                    refreshComments()
                                }
                            }
                        )
                    }
                }
            }

            // Reply indicator
            if (replyingTo != null) {
                Row(
                    modifier = Modifier
                        .fillMaxWidth()
                        .background(colors.surface)
                        .padding(horizontal = 20.dp, vertical = 8.dp),
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    Text(
                        text = "Respondendo a ${replyingTo!!.authorName}",
                        fontSize = 12.sp,
                        color = colors.accent,
                        modifier = Modifier.weight(1f)
                    )
                    IconButton(
                        onClick = { replyingTo = null },
                        modifier = Modifier.size(20.dp)
                    ) {
                        Icon(Icons.Default.Close, null, tint = colors.textSecondary, modifier = Modifier.size(14.dp))
                    }
                }
            }

            // Input
            Box(Modifier.fillMaxWidth().height(0.5.dp).background(colors.divider))
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(horizontal = 16.dp, vertical = 12.dp),
                verticalAlignment = Alignment.CenterVertically
            ) {
                Box(
                    modifier = Modifier
                        .weight(1f)
                        .clip(RoundedCornerShape(20.dp))
                        .background(colors.surface)
                        .padding(horizontal = 16.dp, vertical = 10.dp)
                ) {
                    if (commentText.isEmpty()) {
                        Text("Adicionar coment\u00e1rio...", fontSize = 14.sp, color = colors.textSecondary)
                    }
                    BasicTextField(
                        value = commentText,
                        onValueChange = { commentText = it },
                        textStyle = TextStyle(fontSize = 14.sp, color = colors.textPrimary),
                        cursorBrush = SolidColor(colors.accent),
                        modifier = Modifier.fillMaxWidth()
                    )
                }
                Spacer(Modifier.width(8.dp))
                IconButton(
                    onClick = {
                        if (commentText.isNotBlank() && !isSending) {
                            val text = commentText
                            val parentId = replyingTo?.id
                            commentText = ""
                            replyingTo = null
                            isSending = true
                            scope.launch {
                                ApiClient.addComment(
                                    postId = postId,
                                    text = text,
                                    authorName = TokenManager.displayName ?: "An\u00f4nimo",
                                    parentId = parentId
                                )
                                refreshComments()
                                isSending = false
                            }
                        }
                    },
                    enabled = commentText.isNotBlank() && !isSending,
                    modifier = Modifier.size(36.dp)
                ) {
                    if (isSending) {
                        CircularProgressIndicator(color = colors.accent, strokeWidth = 2.dp, modifier = Modifier.size(18.dp))
                    } else {
                        Icon(
                            Icons.AutoMirrored.Filled.Send,
                            "Enviar",
                            tint = if (commentText.isNotBlank()) colors.accent else colors.textSecondary,
                            modifier = Modifier.size(20.dp)
                        )
                    }
                }
            }
        }
    }
}

@Composable
private fun CommentItem(
    comment: Comment,
    onReply: (Comment) -> Unit,
    onVote: (String, String) -> Unit,
    onDelete: (String) -> Unit,
    isReply: Boolean = false
) {
    val colors = AppTheme.colors
    val currentUserId = TokenManager.userId

    Column(
        modifier = Modifier
            .fillMaxWidth()
            .padding(
                start = if (isReply) 40.dp else 20.dp,
                end = 20.dp,
                top = 12.dp,
                bottom = if (comment.replies.isEmpty() && !isReply) 12.dp else 4.dp
            )
    ) {
        // Author + time
        Row(verticalAlignment = Alignment.CenterVertically) {
            Text(
                text = comment.authorName.ifBlank { "An\u00f4nimo" },
                fontSize = 13.sp,
                fontWeight = FontWeight.SemiBold,
                color = colors.textPrimary
            )
            Spacer(Modifier.width(6.dp))
            Text("\u00B7", fontSize = 12.sp, color = colors.textSecondary)
            Spacer(Modifier.width(6.dp))
            Text(comment.timeAgo, fontSize = 11.sp, color = colors.textSecondary)
        }

        Spacer(Modifier.height(4.dp))

        // Text
        Text(
            text = comment.text,
            fontSize = 14.sp,
            color = colors.textPrimary,
            lineHeight = 19.sp
        )

        Spacer(Modifier.height(6.dp))

        // Actions
        Row(verticalAlignment = Alignment.CenterVertically) {
            // Vote up
            Row(
                verticalAlignment = Alignment.CenterVertically,
                modifier = Modifier
                    .clip(RoundedCornerShape(4.dp))
                    .clickable { onVote(comment.id, "up") }
                    .padding(4.dp)
            ) {
                Icon(Icons.Default.KeyboardArrowUp, "Votar positivo", modifier = Modifier.size(16.dp), tint = colors.textSecondary)
                if (comment.upvotes > 0) {
                    Text("${comment.upvotes}", fontSize = 11.sp, color = colors.textSecondary)
                }
            }

            Spacer(Modifier.width(4.dp))

            // Vote down
            Row(
                verticalAlignment = Alignment.CenterVertically,
                modifier = Modifier
                    .clip(RoundedCornerShape(4.dp))
                    .clickable { onVote(comment.id, "down") }
                    .padding(4.dp)
            ) {
                Icon(Icons.Default.KeyboardArrowDown, "Votar negativo", modifier = Modifier.size(16.dp), tint = colors.textSecondary)
                if (comment.downvotes > 0) {
                    Text("${comment.downvotes}", fontSize = 11.sp, color = colors.textSecondary)
                }
            }

            if (!isReply) {
                Spacer(Modifier.width(12.dp))
                Text(
                    text = "Responder",
                    fontSize = 12.sp,
                    fontWeight = FontWeight.Medium,
                    color = colors.textSecondary,
                    modifier = Modifier
                        .clip(RoundedCornerShape(4.dp))
                        .clickable { onReply(comment) }
                        .padding(4.dp)
                )
            }

            Spacer(Modifier.weight(1f))

            // Delete (if own comment - simplified check)
            IconButton(
                onClick = { onDelete(comment.id) },
                modifier = Modifier.size(24.dp)
            ) {
                Icon(Icons.Default.Delete, "Excluir", modifier = Modifier.size(14.dp), tint = colors.textSecondary.copy(alpha = 0.5f))
            }
        }

        // Replies
        if (comment.replies.isNotEmpty()) {
            Column(modifier = Modifier.animateContentSize()) {
                comment.replies.forEach { reply ->
                    CommentItem(
                        comment = reply,
                        onReply = onReply,
                        onVote = onVote,
                        onDelete = onDelete,
                        isReply = true
                    )
                }
            }
        }

        if (!isReply) {
            Spacer(Modifier.height(8.dp))
            Box(Modifier.fillMaxWidth().height(0.5.dp).background(colors.divider.copy(alpha = 0.5f)))
        }
    }
}
